content/concepts/hole-punching.md

```diff
@@ -0,0 +1,193 @@
+---
+title: Hole Punching
+weight: 3
+---
+
+## Types of nodes
+
+Nodes on the public internet can be divided into two groups: 
+public and non-public. Most nodes are not publicly accessible as 
+routers act as firewalls and allow for NATs. 
+
+To be clear, a non-public node from a standard LAN:
+
+- can dial other public nodes;
+- cannot dial non-public nodes.
+
+Likewise, a public node on the internet:
+
+- can dial other public nodes;
+- cannot dial non-public nodes.
+
+### How can a node dial a non-public node?
+
+Here are a few methods that nodes can use to dial a non-public node:
+
+- UPnP (Universal Plug and Play): A protocol that enables nodes 
+  to discover and connect by automatically opening
+  ports into a firewall.
+- Port forwarding: Manually configuring a port forward on a router.
+- Traversal Using Relay NAT (TURN): A protocol that can traverse
+  a NAT, allowing a client to obtain IP addresses and ports from 
+  relaying.
+
+### Limitations
+
+UPnP automates the process of node discovery and connectivity. Still, 
+it may not be available everywhere, and there can posses the risk
+of establishing connections with untrustworthy nodes as the protocol
+does not enforce authentication or authorization. Manually opening a
+port requires technical expertise and does not enforce 
+authentication or authorization. Using a relay node also requires
+technical expertise. Relaying adds additional latency and is resource
+intensive.
+
+### Possible solution: hole punching
+
+In the case where the other options aren't sufficient, networks can 
+use a technique called hole punching to establish connections with 
+non-public nodes.
+
+Each node connects to an unrestricted third-party server and 
+shares its external and internal address and port information.
+The server temporarily stores the node's information and relays 
+each node's information to the other. Clients can use this 
+information to establish direct connections with each other.
+
+Take two nodes, `A` and `B`, that would like the dial each other:
+
+1. The first packet of both nodes (e.g., in the case of TCP, an SYN) 
+   passes through their respective routers.
+2. The routers add a 5-tuple to their router's state table. 
+3. `Packet A` and `B` "punch holes" into their respective routers' 
+   firewalls.
+4. Both packets arrive at the opposite router.
+5. Once `A`'s packet arrives at `router B`, `router B` checks its state 
+   table and finds a 5-tuple previously added through the packet sent by 
+   node B. 
+6. The routers forward the packets through the "punched holes". 6. It 
+   then forwards the packet to `B`. The same occurs with `B`'s packet; 
+   upon arriving at `router A`, it matches a 5-tuple in `router A`'s state 
+   table and thus forwards the packet to `A`.
+  
+The following use case diagram illustrates the above process.
+
+![](https://i.imgur.com/0k2Zlj3.png)
```

Comment:
Can we not use SVGs here?

Reply:
Yes indeed I guess I've been avoiding to add new assets until https://github.com/libp2p/docs/pull/186 is solved (which would be added thereafter)